108 base on gettext 026

jk-lamb opened the pull request:
# Pull Request

## Description

Adds Gettext 0.26.0 compatibility with a custom backend adapter system:

## Type of Change

<!-- Mark the relevant option with an "x" -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update
- [x] Performance improvement
- [ ] Code refactoring
- [ ] Test improvements
- [ ] CI/CD improvements

## Related Issues

<!-- Link any related issues -->
Fixes #<!-- issue number -->
Closes #108
Related to #107 

## Changes Made

  - Custom Gettext Backend System: Created Kanta.Backend with CachedDB adapter that fetches translations from database/cache instead of PO files at runtime, with fallback support
  - New Architecture: Added backend adapter pattern with Kanta.Backend.Adapter behavior and FallbackBackend for missing translations
  - Migration Updates: Added v04 PostgreSQL and v03 SQLite migrations to support new backend system
  - Removed: Old Kanta.Gettext.Repo module (156 lines) replaced with new adapter system

## Testing

<!-- Describe the tests you ran to verify your changes -->

### Test Environment
- [x] Elixir version: 1.18.4 <!-- e.g., 1.17.0 -->
- [x] OTP version: 27.3.4.1 <!-- e.g., 27 -->
- [x] Phoenix version: 1.7.11 <!-- e.g., 1.7.0 -->
- [x] Database: PostgreSQL 17 <!-- e.g., PostgreSQL 14 or SQLite 3 -->
- [x] Gettext version: 0.26.2 <!-- if using custom fork -->

### Test Cases
- [x] All existing tests pass
- [x] New tests added for new functionality at appropriate levels
- [ ] Manual testing performed

### Test Commands Run
```bash
# List the commands you ran to test
mix test
MIX_ENV=test mix credo
MIX_ENV=test mix dialyzer
```

## Documentation

- [x] Updated README.md (if applicable)
- [ ] Updated documentation comments (with examples for new features)
- [x] Updated CHANGELOG.md (if applicable)

## Code Quality

- [x] Code follows the existing style conventions
- [x] Self-review of the code has been performed
- [x] Code has been commented, particularly in hard-to-understand areas
- [x] No new linting warnings introduced
- [x] No new Dialyzer warnings introduced

## Backward Compatibility

- [ ] This change is backward compatible
- [x] This change includes breaking changes (please describe below)
- [x] Migration guide provided for breaking changes

### Breaking Changes

Adds Gettext 0.26.0 compatibility with a custom backend adapter system: Will not work with Gettextt < 0.26

## Performance Impact

- [x] No performance impact
- [ ] Performance improvement
- [ ] Potential performance regression (please describe)

### Performance Notes
<!-- Describe any performance considerations -->

## Translation Management Impact

- [ ] No impact on existing translations
- [x] Affects translation extraction process
- [ ] Affects translation storage/retrieval
- [ ] Affects Kanta UI/dashboard
- [ ] Affects plugin system
- [ ] Database schema changes

### Translation Impact Notes
<!-- Describe how this change affects translation workflows -->

ettext 0.26 Migration:
  - Migrates from experimental Gettext branch to official 0.26 release with backend support
  - Replaces Kanta.Gettext.Repo (156 lines removed) with new Kanta.Backend.Adapter.CachedDB backend adapter
  - Introduces fallback mechanism to PO files via Kanta.Backend.GettextFallback

  Translation Storage/Retrieval:
  - New backend adapter fetches translations from database/cache at runtime instead of compiled PO files
  - Maintains existing caching layer for performance
  - All existing translations remain intact but retrieval mechanism changed

  Extraction Process:
  - Added Kanta.Utils.GettextRecompiler for managing recompilation during extraction
  - New flag file system (priv/kanta/.fallback_recompiled) tracks extraction state
  - Improved extraction handling with ModuleFolder utilities

  Database Schema Changes:
  - PostgreSQL v3 → v4 migration: Adds default context, updates null context_id values
  - SQLite v2 → v3 migration: Adds default context, updates null context_id values
  - Migration required for all existing installations

## Security Considerations

- [x] No security impact
- [ ] Security improvement
- [ ] Potential security impact (please describe)

## Additional Notes

<!-- Any additional information that reviewers should know -->
<!-- For UI changes, please include before/after screenshots -->
<!-- For new plugins, include setup/configuration examples -->

Plugin System Impact:
  - Internal API changes from Kanta.Gettext.Repo to backend adapter may require plugin updates
  - Plugins relying on old repo interface need migration to new backend adapter API
  - DeepL, POWriter, and KantaSync plugins should be tested for compatibility

## Screenshots/Examples

<!-- If applicable, add screenshots of UI changes or code examples -->

```elixir
# Example usage of new translation management feature
# For UI changes, show before/after screenshots
# For API changes, show usage example:

# Example: New translation extraction feature
Kanta.extract_translations_from_po_files(domain: "errors")

# Example: New plugin integration
defmodule MyApp.CustomTranslationPlugin do
  use Kanta.Plugin
  # implementation
end
```

## Checklist

- [x] I have read the [Contributing Guidelines](CONTRIBUTING.md)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published

## Reviewer Notes

<!-- Any specific areas you'd like reviewers to focus on -->

---

<!-- Thank you for contributing to Kanta! -->

arturz committed:
Update gettext version to 0.26.0+

arturz committed:
Create Kanta gettext adapter

arturz committed:
Create custom gettext module

arturz committed:
Remove Gettext repo module

arturz committed:
Add elixirc_paths and postgrex

arturz committed:
Update Gettext configuration instruction

arturz committed:
Configure test section in config file

arturz committed:
Handle interpolations in messages from DB or cache

arturz committed:
Add create_locale function

arturz committed:
Set up tests helpers

arturz committed:
Remove TODO

arturz committed:
Pass bindings into adapter functions

arturz committed:
Create fixtures for tests

arturz committed:
Create basic test cases for Kanta.Gettext module

arturz committed:
Add multi_messages fixtures

arturz committed:
Create extractor test

arturz committed:
Create macros test

arturz committed:
Create modified gettext test

arturz committed:
Remove unused alias

arturz committed:
Change typespec to satisfy both 1.14 and 1.15+ formatters

arturz committed:
Update GitHub workflow

arturz committed:
Delegate all public functions in Gettext module

arturz committed:
Update warnings to use Kanta.Gettext

jk-lamb committed:
Merge pull request #113 from curiosum-dev/develop

Develop

jk-lamb committed:
Gettext 0.26.0 compatibility

- Add Gettext.Backend implementation (Gettext 0.26.2) for Kanta.
- Kanta Cached DB
- Add Fallback Gettext backend for PO file checks
- Gettext.Recompiler util module to support one time recompilation after mix gettext.extract

Tests:
- Add migration(v4) for test repo
- Re-implemented DataCase helper (manual checkout)
- Add tests for the Backend and Adapter

jk-lamb committed:
Fix: non async tests for DB-related operation.

jk-lamb committed:
Minor fixes

- Remove Logger.debug(...)
- add removing trailing "." to safe_folder_name (for Windows)

jk-lamb committed:
Update README.md

jk-lamb committed:
Merge branch 'develop' into 108-base-on-gettext-026

jk-lamb committed:
Update mix.exs: Relax gettext dependency constraint (0.26.*)

jk-lamb committed:
Update README.md

jk-lamb committed:
Update CHANGELOG.md

jk-lamb committed:
Add missing @moduledoc annotations

The changes add missing @moduledoc attributes to various modules,
providing clear documentation about their purpose.

jk-lamb committed:
Update README.md

jk-lamb committed:
Add migration step to GitHub Actions database setup

mix ecto.drop && mix ecto.create && mix ecto.migrate

arturz commented on CHANGELOG.md:
```diff
@@ -6,6 +6,8 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 ## Unreleased
 
+- Adds Gettext 0.26.0 compatibility with a custom backend adapter system
```
Add information about breaking change (the necessity to update the Gettext definition).

arturz commented on lib/kanta/backend/adapter/cached_db.ex:
```diff
@@ -0,0 +1,144 @@
+          missing_bindings: missing_bindings
+        })
+
+        :not_found
```
In the codebase, we return either `{:ok, value}` or `{:error, reason}`, so it should be `{:error, :not_found}`.

arturz commented on lib/kanta/backend/adapter/cached_db.ex:
```diff
@@ -0,0 +1,144 @@
+defmodule Kanta.Backend.Adapter.KantaCachedDB do
```
Is the `Kanta` prefix necessary in `KantaCachedDB`? Filename doesn't reflect this

arturz commented on test/kanta/backend/adapter/cached_db_test.exs:
```diff
@@ -0,0 +1,236 @@
+
+  setup do
+    # Clear the cache before each test
+    Kanta.Cache.delete_all()
```
Sidenote: please make sure that the cache is also updated/cleared when manipulating through UI

jk-lamb committed:
Address PR review feedback

- Add breaking change notice to CHANGELOG.md explaining Gettext 0.26.0 migration requirements
- Fix apply_bindings return type from :not_found to {:error, :not_found} to match codebase conventions
- Rename KantaCachedDB to CachedDB module (remove redundant Kanta prefix)
- Update all references to use new module name

jk-lamb committed:
Address PR review feedback

- Add breaking change notice to CHANGELOG.md explaining Gettext 0.26.0
migration requirements - Fix apply_bindings return type from :not_found
to {:error, :not_found} to match codebase conventions - Rename
KantaCachedDB to CachedDB module (remove redundant Kanta prefix) -
Update all references to use new module name

jk-lamb committed:
Move recompilation flags to _build directory

Relocates GettextRecompiler flag files from priv/kanta/ to environment-specific build directories using Mix.Project.build_path(). This prevents confusion about whether temporary build artifacts should be committed to version control, as _build/ is already in .gitignore.

The flag files now reside in:
- _build/{env}/kanta_recompile/.gettext_recompiled
- _build/{env}/kanta_recompile/.fallback_recompiled

jk-lamb committed:
Merge branch '108-base-on-gettext-026' of https://github.com/curiosum-dev/kanta into 108-base-on-gettext-026

# Conflicts:
#	lib/kanta/backend.ex

jk-lamb committed:
Rename cached_db test file to match module name

jk-lamb committed:
Fix line break in cached DB test
